### PR TITLE
Plugins: Angular deprecation: Detect Angular plugins and expose in API

### PR DESCRIPTION
**Pull request by xnyo**
**What is this feature?**

Exposes `angular` in `/api/plugins` which is `true` if a plugin is using Angular.

Code for angular deprecation detection is taken from the plugin validator's `legacyplatform` pass:
https://github.com/grafana/plugin-validator/tree/5f95b4e68d9335a178de15ced1177d2e119bbdca/pkg/analysis/passes/legacyplatform

**Why do we need this feature?**

First step to display Angular status for installed plugins, as part of Angular deprecation.

**Who is this feature for?**

Grafana users.

**Which issue(s) does this PR fix?**:

Related to:
- https://github.com/grafana/grafana/issues/64143
- https://github.com/grafana/grafana/issues/65340

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by tolzhabayev** [comment]
@xnyo take a look at this patterns https://github.com/grafana/grafana-plugins-platform-detect-angular/blob/main/patterns.mjs we should make them and the ones in https://github.com/grafana/plugin-validator/tree/5f95b4e68d9335a178de15ced1177d2e119bbdca/pkg/analysis/passes/legacyplatform the same (merge the lists).

File: pkg/plugins/localfiles.go [hunk 1]
```diff
@@ -137,3 +139,62 @@ func (p *LocalFile) Close() error {
 	p.f = nil
 	return nil
 }
+
+type inMemoryFS struct {
+	files map[string][]byte
+}
+
+func NewInMemoryFS(files map[string][]byte) FS {
```
Comment (andresmgot) [inline pkg/plugins/localfiles.go:147]: I would also move this to a test-only file :)

**Review by andresmgot — COMMENTED** [review]
Neat! It would be easy for the frontend to react to the info returned by this (not sure what we want to show though). Do we know any plugin that is using Angular so I can try it?

**Comment by xnyo** [comment]
@andresmgot some Angular plugins:
https://grafana.com/grafana/plugins/grafana-worldmap-panel/
https://grafana.com/grafana/plugins/singlestat/

**Comment by leventebalogh** [comment]
Nice one @xnyo! 👏 🎉 

Would it make sense to run this checker (or an identical one) on all the available plugins in the catalog and update the GCOM database, so all existing plugins have this information by default? I guess the benefit of that would be that users also see if a plugin is Angular-based without installing them, or we can even filter them out / highlight them at some point. 
(Like this they would only see if a plugin is Angular after they have installed them and the bundle is downloaded if I am correct?)

**Comment by andresmgot** [comment]
> Would it make sense to run this checker (or an identical one) on all the available plugins in the catalog and update the GCOM database, so all existing plugins have this information by default? I guess the benefit of that would be that users also see if a plugin is Angular-based without installing them, or we can even filter them out / highlight them at some point. Like this they would only see if a plugin is Angular after they have installed them and the bundle is downloaded if I am correct? (The recognition came from @jackw yesterday while we were having a chat.)

That makes sense, even though I guess we already have the list of Angular plugins so we can do that manually? (@sympatheticmoose or @tolzhabayev should know). In any case, that's a parallel thing, I think we can go ahead with this approach.

**Comment by sympatheticmoose** [comment]
Sorry just dropping in and havent had a proper chance to review, so just on the last question from @andresmgot Yup - we've got a list of plugins where the _latest_ version is running Angular, which I guess is all we'd need for this usage? (Also agree these are parallel concerns)

**Comment by xnyo** [comment]
@sympatheticmoose This PR inspects the source code of any locally installed plugin (even private ones, unsigned, and ones that are not in the catalog). It's only a change in the backend for now, I haven't done anything that's visible by the end user yet.

If we want to extend this to the plugins displayed in the catalog page (and I personally think we should, like you and @leventebalogh suggested), we would need such list and add this info to the GCOM (grafana.com) API.

If the current list contains detection only for the latest version of all public plugins, we can extend that to all versions by running the angular detection script on the CDN, since it is populated with all plugin versions ever, not just the latest :smile:

**Comment by tolzhabayev** [comment]
@xnyo @sympatheticmoose I think this is a great spike and possibility to detect angular plugins that are not part of our catalog - private ones for example.

I would suggest not moving this into the core yet, but rather have it as a future reference and refine what exactly we want to do in which UI / Area etc. as it might have influence on the backend implementation.

One thing we should also take a look at is if we want / can make this patterns more dynamic so that we can update the list without updating the grafana version? And maybe fallback to the current list in situations that are not possible

**Comment by torkelo** [comment]
@xnyo any progress on this? I need this for more warnings like https://github.com/grafana/grafana/pull/67891 but for data sources

**Comment by xnyo** [comment]
@torkelo I have fixed the tests and opened this PR so it will be merged soon. I have two questions:
1. This PR adds an `angular` property in `/api/plugins` as it was meant to be used in the plugins catalog page. Do you need it somewhere else as well, for example in `grafanaBootData` ?
2. Do you need this backported to 10.0.0 as well?

File: pkg/api/dtos/plugins.go [hunk 2]
```diff
@@ -45,6 +45,7 @@ type PluginListItem struct {
 	SignatureType plugins.SignatureType   `json:"signatureType"`
 	SignatureOrg  string                  `json:"signatureOrg"`
 	AccessControl accesscontrol.Metadata  `json:"accessControl,omitempty"`
+	Angular       bool                    `json:"angular"`
```
Comment (torkelo) [inline pkg/api/dtos/plugins.go:48]:
```suggestion
	AngularDetected       bool                    `json:"angularDetected"`
```

Just to make it more clear what this prop means

**Review by torkelo — COMMENTED** [review]
It would great if the property was included in the datasource & panel plugin meta object (from the frontend).

**Review by torkelo — COMMENTED** [review]
Looks good!

**Review by marefr — APPROVED** [review]
LGTM great work

**Pull request merged**